access/resource_table_acl.go

```diff
@@ -0,0 +1,351 @@
+	for result.Scan(&principal, &actionType, &objType, &objectKey) {
+		if !strings.EqualFold(objectType, objType) {
+			continue
+		}
+		if !strings.EqualFold(objectKey, key) {
+			continue
+		}
+		target := ta.Grants
+		var isDeny = strings.HasPrefix(actionType, "DENIED_")
+		if isDeny {
+			actionType = strings.Replace(actionType, "DENIED_", "", 1)
+			target = ta.Denies
+		}
+		var privileges *[]string
+		for i, tp := range target {
+			if tp.Principal == principal {
+				if isDeny {
+					privileges = &ta.Denies[i].Privileges
+				} else {
+					privileges = &ta.Grants[i].Privileges
+				}
+			}
+		}
+		if privileges == nil {
+			tablePerms := TablePermissions{
+				Principal:  principal,
+				Privileges: []string{},
+			}
+			if isDeny {
+				ta.Denies = append(ta.Denies, tablePerms)
+				privileges = &ta.Denies[len(ta.Denies)-1].Privileges
+			} else {
+				ta.Grants = append(ta.Grants, tablePerms)
+				privileges = &ta.Grants[len(ta.Grants)-1].Privileges
+			}
+		}
+		*privileges = append(*privileges, actionType)
+	}
+	return nil
+}
+
+func (ta *TableACL) revoke() error {
+	existing, err := loadTableACL(ta.ID())
+	if err != nil {
+		return err
+	}
+	existing.exec = ta.exec
+	existing.ClusterID = ta.ClusterID
+	if err = existing.read(); err != nil {
+		return err
+	}
+	for _, tps := range existing.permissions() {
+		for _, priv := range tps {
+			if err = ta.apply(func(objType, key string) string {
+				return fmt.Sprintf("REVOKE ALL PRIVILEGES ON %s %s FROM `%s`",
+					objType, key, priv.Principal)
+			}); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (ta *TableACL) enforce() (err error) {
+	if err = ta.revoke(); err != nil {
+		return err
+	}
+	for action, grantsOrDenies := range ta.permissions() {
+		for _, grant := range grantsOrDenies {
+			if err = ta.apply(func(objType, key string) string {
+				privileges := strings.Join(grant.Privileges, ", ")
+				return fmt.Sprintf("%s %s ON %s %s TO `%s`",
+					action, privileges, objType, key, grant.Principal)
+			}); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (ta *TableACL) apply(qb func(objType, key string) string) error {
+	sqlQuery := qb(ta.TypeAndKey())
+	log.Printf("[INFO] Executing SQL: %s", sqlQuery)
+	r := ta.exec.Execute(ta.ClusterID, "sql", sqlQuery)
+	return r.Err()
+}
+
+func (ta *TableACL) initCluster(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) (err error) {
+	clustersAPI := compute.NewClustersAPI(ctx, c)
+	if ci, ok := d.GetOk("cluster_id"); ok {
+		ta.ClusterID = ci.(string)
+	} else {
+		ta.ClusterID, err = ta.getOrCreateCluster(clustersAPI)
+		if err != nil {
+			return
+		}
+	}
+	clusterInfo, err := clustersAPI.StartAndGetInfo(ta.ClusterID)
+	if err != nil {
+		return
+	}
+	if v, ok := clusterInfo.SparkConf["spark.databricks.acl.dfAclsEnabled"]; !ok || v != "true" {
+		err = fmt.Errorf("cluster_id: not a High-Concurrency cluster: %s (%s)",
+			clusterInfo.ClusterName, clusterInfo.ClusterID)
```

Comment:
What does high concurrency have to do with table ACLs?

Reply:
You won't believe it - "when a high number of users use the same cluster concurrently, we can provide an enhanced security for that cluster using table acls". E.g. it is synonymous.